scripts/release/audit_rosdistro.py

```diff
@@ -0,0 +1,135 @@
+# limitations under the License.
+
+
+import argparse
+import copy
+import sys
+
+from ros_buildfarm.argument import add_argument_cache_dir
+from ros_buildfarm.argument import add_argument_config_url
+from ros_buildfarm.argument import add_argument_return_zero
+from ros_buildfarm.argument import add_argument_rosdistro_name
+from ros_buildfarm.config import get_index as get_config_index
+from ros_buildfarm.config import get_release_build_files
+from ros_buildfarm.release_job import partition_packages
+from ros_buildfarm.status_page import get_jenkins_job_urls
+from rosdistro import get_distribution_cache
+from rosdistro import get_distribution_file
+from rosdistro import get_index
+
+
+def main(argv=sys.argv[1:]):
+    parser = argparse.ArgumentParser(
+        description='Audit the rosdistro for packages failing to build.')
+    add_argument_config_url(parser)
+    add_argument_rosdistro_name(parser)
+    add_argument_cache_dir(parser, '/tmp/package_repo_cache')
+    add_argument_return_zero(parser)
+    args = parser.parse_args(argv)
+
+    recommended_action_count = run_audit(args.config_url, args.rosdistro_name, args.cache_dir)
+    return 1 if recommended_action_count and not args.return_zero else 0
+
+
+def run_audit(config_url, rosdistro_name, cache_dir):
+    config = get_config_index(config_url)
+    index = get_index(config.rosdistro_index_url)
+    dist_file = get_distribution_file(index, rosdistro_name)
+    dist_cache = get_distribution_cache(index, rosdistro_name)
+    build_files = get_release_build_files(config, rosdistro_name)
+    missing_packages = {}
+    for bf_name, bf_value in build_files.items():
+        missing_packages[bf_name] = copy.deepcopy(bf_value.targets)
+        for target in bf_value.get_targets_list():
+            all_pkgs, missing_pkgs = partition_packages(
+                config_url, rosdistro_name, bf_name, target, cache_dir,
+                deduplicate_dependencies=True, dist_cache=dist_cache)
+            missing_packages[bf_name][target] = missing_pkgs
+            if 'all' in missing_packages[bf_name]:
+                missing_packages[bf_name]['all'] &= missing_pkgs
+            else:
+                missing_packages[bf_name]['all'] = missing_pkgs
+
+            if 'all' in missing_packages:
+                missing_packages['all'] &= missing_pkgs
+            else:
+                missing_packages['all'] = missing_pkgs
+
+    recommended_actions = len(missing_packages['all'])
+    print('# Sync preparation report for %s' % rosdistro_name)
+    print('Prepared for configuration: %s' % config_url)
+    print('Prepared for rosdistro index: %s' % config.rosdistro_index_url)
+
+    print('## Packages failing on all platforms\n Release should be rolled back:')
+    for mp in sorted(missing_packages['all']):
+        print(' - %s ' % mp)
+
+    def get_package_repository_link(dist_file, pkg_name):
+        """Return the best guess of the url for filing a ticket against the package."""
```

Comment:
It doesn't seem like we necessarily have cached package.xml data at this stage but if we did and the package specified an issue url this could be even more precise.